pkg/ccl/streamingccl/streamproducer/event_stream.go

```diff
@@ -255,97 +199,85 @@ func (s *eventStream) Close(ctx context.Context) {
 		s.rf.Close()
 	}
 	s.acc.Close(ctx)
-	if s.doneChan != nil {
-		close(s.doneChan)
-	}
-	if err := s.streamGroup.Wait(); err != nil {
-		// Note: error in close is normal; we expect to be terminated with context canceled.
-		log.Errorf(ctx, "partition stream %d terminated with error %v", s.streamID, err)
-	}
-	if s.sp != nil {
-		s.sp.Finish()
-	}
 }
 
 func (s *eventStream) onValue(ctx context.Context, value *kvpb.RangeFeedValue) {
-	select {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{RangeFeedEvent: &kvpb.RangeFeedEvent{Val: value}}:
-		log.VInfof(ctx, 1, "onValue: %s@%s", value.Key, value.Value.Timestamp)
-	}
-}
-
-func (s *eventStream) onCheckpoint(ctx context.Context, checkpoint *kvpb.RangeFeedCheckpoint) {
-	select {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{RangeFeedEvent: &kvpb.RangeFeedEvent{Checkpoint: checkpoint}}:
-		log.VInfof(ctx, 1, "onCheckpoint: %s@%s", checkpoint.Span, checkpoint.ResolvedTS)
-	}
-}
-
-func (s *eventStream) onInitialScanSpanCompleted(ctx context.Context, sp roachpb.Span) error {
-	checkpoint := kvpb.RangeFeedCheckpoint{
-		Span:       sp,
-		ResolvedTS: s.spec.InitialScanTimestamp,
-	}
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case s.eventsCh <- kvcoord.RangeFeedMessage{
-		RangeFeedEvent: &kvpb.RangeFeedEvent{Checkpoint: &checkpoint},
-	}:
-		log.VInfof(ctx, 1, "onSpanCompleted: %s@%s", checkpoint.Span, checkpoint.ResolvedTS)
-		return nil
+	s.seb.addKV(roachpb.KeyValue{Key: value.Key, Value: value.Value})
+	if err := s.maybeFlushBatch(ctx); err != nil {
+		s.errCh <- err
```

Comment:
I'm not 100% convinced this is correct. I _nearly_ convinced myself because the errCh channel is buffered and the callbacks aren't called concurrently. 

But, then I realized that last bit isn't true during the initial scan which will call the onValue callback concurrently.  That poses a problem not just for this errCh, but it also makes me question whether the initial scan will be a problem for the stream_event_batcher which has no mutex.

If I'm right that this is a problem, it's a little concerning that we don't have a single unit test that failed here.

Reply:
Ah, yeah, I took the comment in rangefeed.go that said `callbacks... are called in said async task in a single thread.` at face value, but looks like initial scans violate that comment if concurrency is enabled. I added a mutex to add (that is only used during initial scan). The rest of them I think are safe, as long as they `select` to discard errors if one is already in the ch.